serialization_v2/serialization.py

```diff
@@ -0,0 +1,1280 @@
+        # and thus the generated code might fail.
+        # So add all other paths here which we currently have.
+        # (While testing this, this was sisyphus + returnn + recipes,
+        #  but returnn is excluded below.)
+        sys_paths = set(_get_base_sys_path_list())
+        extra_sys_paths = [p for p in sys.path if p not in sys_paths]
+
+        # Handle ExternalImports
+        extra_sys_paths += [
+            item.import_path
+            for item in tree.flatten(config) + tree.flatten(post_config)
+            if isinstance(item, ExternalImport)
+        ]
+
+        python_prolog_code = unparse_python(self.python_prolog, hash_full_python_code=True)
+        python_epilog_code = unparse_python(self.python_epilog, hash_full_python_code=True)
+        serialized = serialize_config(
+            config,
+            post_config,
+            # Of course RETURNN knows about itself, no need to add to sys.path.
+            # Also, we don't want to force the current RETURNN here,
+            # but allow the config to be used with any other RETURNN version.
+            known_modules={"returnn"},
+            extra_sys_paths=extra_sys_paths,
+            # instanciate_delayed should already have handled it (e.g. Path),
+            # or if it has not, then we want to keep it as it is (e.g. PtCheckpoint),
+            # but without dependencies.
+            sis_path_handling=SisPathHandling.AS_STRING,
+        )
+        return "\n\n".join(
+            part
+            for part in [
+                "#!returnn/rnn.py",
+                python_prolog_code,
+                serialized.as_serialized_code(),
+                python_epilog_code,
+                "# -*- mode: python; tab-width: 4 -*-\n",
+            ]
+            if part.strip()
+        )
+
+
+@dataclass
+class SerializedConfig:
+    code_list: List[PyCode]
+
+    def as_serialization_collection(self) -> SerializerCollection:
+        """as serialization Collection"""
+        return SerializerCollection(self.code_list)
+
+    def as_serialized_code(self) -> str:
+        """as serialized code"""
+        return "".join(code.py_code for code in self.code_list)
+
+
+class _Serializer:
+    def __init__(
+        self,
+        config: Dict[str, Any],
+        post_config: Optional[Dict[str, Any]] = None,
+        known_modules: Collection[str] = (),
+        sis_path_handling: SisPathHandling = None,
+    ):
+        self.config = config.copy()
+        self.post_config = post_config.copy() if post_config else {}
+        self.assignments_dict_by_value_ref: Dict[_Ref, PyCode] = {}  # value ref -> code
+        self.assignments_dict_by_name: Dict[str, PyCode] = {}  # var name -> code
+        self.assignments_dict_by_idx: Dict[int, PyCode] = {}  # idx -> code
+        self.assignments_dict_by_value_by_type: Dict[type, Dict[Any, PyCode]] = {
+            # TODO: is this entry needed for correctness?
+            # Dim: {},
```

Comment:
I think this is needed, because the `Dim` serialization is actually simplified, and drops some of the deeper structures, including the `same_as` stuff, and then two dim objects which are equal but not the same instance would not be equal anymore after the serialization.

Maybe you can add such code below:
```python
if "returnn" in sys.modules:
    from returnn.tensor import Dim

    self.assignments_dict_by_value_by_type[Dim] = {}
```